6c information gathering (#1589)
OWASP/owasp-mstg: Document/0x06c-Reverse-Engineering-and-Tampering.md

To learn more, please refer to the [r2frida wiki](https://github.com/enovella/r2frida-wiki/blob/master/README.md "r2frida Wiki").

> ##### Information Gathering

Reviewer: Can you move this section to line 716 instead? We should describe first how to get the names of classes/methods before we are describing "Method hooking".

> In previous section we learned about using radare2 to collect runtime information of a running process. Radare2 is internally using Frida (r2frida) to obtain process information. In this section we will learn how to use Frida directly to gather process related information.

Reviewer: When moving this part, please remove the first sentence.

##### Getting Loaded Classes and their Methods

> In Frida REPL Objective-C runtime related information can be accessed using `ObjC` command. Within `ObjC` command, `enumerateLoadedClasses` function lists the loaded classes for a given application.

Reviewer: ```suggestion
In the Frida REPL Objective-C runtime the `ObjC` command can be used to access information within the running app. Within the `ObjC` command the function `enumerateLoadedClasses` lists the loaded classes for a given application.
```

```
$ frida -U -f com.iOweAss

> [iPhone::com.iOweAss]-> ObjC.enumerateLoadedClasses()

Reviewer: Which app is this, iOweAss?

Author: It is a crackme app I [wrote](https://github.com/su-vikas/crackmes/tree/master/ios), and was present on my device while writing this. The uncrackable level 1 dont have enough classes to show this example. Is it fine?

Reviewer: Got it. Can you please remove the "Ass" :-)

Reviewer: Is this app publicly available?

Author: haha sure. Yes, I linked above, it is here:  https://github.com/su-vikas/crackmes

Author: Fixed!

{
    "/System/Library/Frameworks/CoreFoundation.framework/CoreFoundation": [
        "__NSBlockVariable__",
        "__NSGlobalBlock__",
        "__NSFinalizingBlock__",
        "__NSAutoBlock__",
        "__NSMallocBlock__",
        "__NSStackBlock__"
    ],
    "/private/var/containers/Bundle/Application/F390A491-3524-40EA-B3F8-6C1FA105A23A/iOweAss.app/iOweAss": [
        "JailbreakDetection",
        "CriticalLogic",
        "ViewController",
        "AppDelegate"
    ]
}

```

Using `ObjC.classes.<classname>.$ownMethods` the methods declared in each class can be listed.

```
[iPhone::com.iOweAss]-> ObjC.classes.JailbreakDetection.$ownMethods
[
    "+ isJailbroken"
]

[iPhone::com.iOweAss]-> ObjC.classes.CriticalLogic.$ownMethods
[
    "+ doSha256:",
    "- a:",
    "- AES128Operation:data:key:iv:",
    "- coreLogic",
    "- bat",
    "- b:",
    "- hexString:"
]
```

##### Getting Loaded Libraries

> In Frida REPL process related information ca be obtained using `Process` command. Within `Process` command, `enumerateModules` function lists the libraries loaded into process memory.

Reviewer: ```suggestion
In Frida REPL process related information can be obtained using the `Process` command. Within the `Process` command the function `enumerateModules` lists the libraries loaded into the process memory.
```

```
[iPhone::com.iOweAss]-> Process.enumerateModules()
[
    {
        "base": "0x10008c000",
        "name": "iOweAss",
        "path": "/private/var/containers/Bundle/Application/F390A491-3524-40EA-B3F8-6C1FA105A23A/iOweAss.app/iOweAss",
        "size": 49152
    },
    {
        "base": "0x1a1c82000",
        "name": "Foundation",
        "path": "/System/Library/Frameworks/Foundation.framework/Foundation",
        "size": 2859008
    },
    {
        "base": "0x1a16f4000",
        "name": "libobjc.A.dylib",
        "path": "/usr/lib/libobjc.A.dylib",
        "size": 200704
    },

    ...
```

Similarly, information related to various threads can be obtained.

```
Process.enumerateThreads()
[
    {
        "context": {
            "fp": "0x16fd72b90",
            "lr": "0x1a17a7aa0",

>             "pc": "0x1a17a8634",

Reviewer: Can you please remove all lines from here till "x9"? To make the output more readable.
Just replace it with "..."

Author: Done

            "sp": "0x16fd72b40",
            "x0": "0x16fd72ca0",
            "x1": "0x7000806",
            "x10": "0xcdcefcf335a90034",
            "x11": "0xb2f482679a",
            "x12": "0x16e3600",
            "x13": "0x963688",
            "x14": "0x0",
            "x15": "0x0",
            "x16": "0xffffffffffffffe1",
            "x17": "0x0",
            "x18": "0x0",
            "x19": "0x0",
            "x2": "0x0",
            "x20": "0xffffffff",
            "x21": "0x2a03",
            "x22": "0xc00",
            "x23": "0x16fd72ca0",
            "x24": "0x7000806",
            "x25": "0x0",
            "x26": "0x7000806",
            "x27": "0xc00",
            "x28": "0x1",
            "x3": "0xc00",
            "x4": "0x2a03",
            "x5": "0xffffffff",
            "x6": "0x0",
            "x7": "0x0",
            "x8": "0xfffffbbf",
            "x9": "0x7000906"
        },
        "id": 1287,
        "state": "waiting"
    },

    ...
```

> `Process` command exposes multiple functions which can be explored as per needs, some useful functions being `findModuleByAddress`, `findModuleByName`, `enumerateRanges` etc.

Reviewer: ```suggestion
The `Process` command exposes multiple functions which can be explored as per needs. Some useful functions are `findModuleByAddress`, `findModuleByName` and `enumerateRanges` besides others.
```
